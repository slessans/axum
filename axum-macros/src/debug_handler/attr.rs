use std::collections::HashMap;

use syn::{parse::Parse, punctuated::Punctuated, Token, Type};

struct GenericArgSpecializationAttr {
    arg_name: syn::Ident,
    specialization_ty: syn::Type,
}

impl Parse for GenericArgSpecializationAttr {
    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
        let arg_name = input.parse()?;
        input.parse::<syn::Token![=]>()?;
        let specialization_ty = input.parse()?;
        Ok(Self {
            arg_name,
            specialization_ty,
        })
    }
}

struct SpecializationsAttr {
    specializations: HashMap<syn::Ident, Vec<syn::Type>>,
}

impl Parse for SpecializationsAttr {
    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
        let specialization_groups: Punctuated<
            Punctuated<GenericArgSpecializationAttr, Token![,]>,
            Token![;],
        > = Punctuated::parse_separated_nonempty_with(input, Punctuated::parse_separated_nonempty)?;
        let mut specializations: HashMap<syn::Ident, Vec<syn::Type>> = HashMap::new();
        for spec_group in specialization_groups {
            let arg_name = match spec_group.first() {
                Some(spec) => spec.arg_name.clone(),
                None => continue, // this should never happen due to parse_nonempty
            };
            if specializations.contains_key(&arg_name) {
                return Err(syn::Error::new(
                    arg_name.span(),
                    "Duplicate argument name specified. Each argument can only appear in one group. Groups are separated by semicolons. (e.g. `T = Foo, T = Bar; U = i64` is valid, but `T = Foo, T = Bar; T = i64` is not).",
                ));
            }
            let mut group_specs = Vec::new();
            for spec in spec_group {
                if spec.arg_name != arg_name {
                    return Err(syn::Error::new(
                        spec.arg_name.span(),
                        "All argument names in a group must match. Groups should be separated by `;`. (e.g. `T = Foo, T = Bar; U = i64` is valid, but `T = Foo, U = Bar` is not).",
                    ));
                }
                group_specs.push(spec.specialization_ty);
            }
            specializations.insert(arg_name, group_specs);
        }
        Ok(Self { specializations })
    }
}

pub(crate) struct Attrs {
    body_ty: Type,
    with_tys: Option<SpecializationsAttr>,
}

impl Parse for Attrs {
    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
        let mut body_ty = None;
        let mut with_tys = None;

        while !input.is_empty() {
            let ident = input.parse::<syn::Ident>()?;
            if ident == "body" {
                input.parse::<Token![=]>()?;
                body_ty = Some(input.parse()?);
            } else if ident == "with" {
                let content;
                syn::parenthesized!(content in input);
                with_tys = Some(content.parse()?);
            } else {
                return Err(syn::Error::new_spanned(ident, "unknown argument"));
            }

            let _ = input.parse::<Token![,]>();
        }

        let body_ty = body_ty.unwrap_or_else(|| syn::parse_quote!(axum::body::Body));

        Ok(Self { body_ty, with_tys })
    }
}

impl Attrs {
    pub(crate) fn body_ty(&self) -> &Type {
        &self.body_ty
    }
    pub(crate) fn specializations(&self) -> Option<&HashMap<syn::Ident, Vec<syn::Type>>> {
        self.with_tys.as_ref().map(|f| &f.specializations)
    }
}
