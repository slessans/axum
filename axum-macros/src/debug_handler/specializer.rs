use std::collections::{HashMap, HashSet};

use itertools::Itertools;
use proc_macro2::TokenStream;
use syn::{
    parse_quote,
    visit::{self, Visit},
    visit_mut::{self, VisitMut},
    GenericParam, ItemFn,
};

use quote::quote;

use super::attr::Attrs;

struct GenericFinder<'a> {
    found_idents: HashSet<&'a syn::Ident>,
    generic_param_set: HashSet<&'a syn::Ident>,
}

impl<'ast, 'a> Visit<'ast> for GenericFinder<'a> {
    fn visit_ident(&mut self, ident: &'ast syn::Ident) {
        if let Some(a) = self.generic_param_set.get(ident) {
            self.found_idents.insert(a);
        }
        // Delegate to the default impl to visit nested expressions.
        visit::visit_ident(self, ident);
    }
}

struct TypeSpecializer<'a> {
    specializations: &'a HashMap<&'a syn::Ident, &'a syn::Type>,
}

impl<'a> VisitMut for TypeSpecializer<'a> {
    fn visit_ident_mut(&mut self, ident: &mut syn::Ident) {
        if let Some(specialized) = self.specializations.get(ident) {
            *ident = parse_quote!(#specialized);
        }
        // Delegate to the default impl to visit nested expressions.
        visit_mut::visit_ident_mut(self, ident);
    }
}
pub(crate) struct Specializer {
    generic_params: Vec<syn::Ident>,
    specializations: HashMap<syn::Ident, Vec<syn::Type>>,
}

impl Specializer {
    pub(crate) fn new(attr: &Attrs, item_fn: &ItemFn) -> Result<Self, syn::Error> {
        let specializations = attr.compute_specializations();

        let generic_params = item_fn
            .sig
            .generics
            .params
            .iter()
            .enumerate()
            .map(|(_idx, param)| match param {
                GenericParam::Type(t) => Ok(t.ident.clone()),
                _ => Err(syn::Error::new_spanned(
                    param,
                    "Only type params are supported by `#[axum_macros::debug_handler]`.",
                )),
            })
            .collect::<Result<Vec<_>, _>>()?;

        // let a = attr.wi

        for param in &generic_params {
            if !specializations.contains_key(param) {
                return Err(
                    syn::Error::new_spanned(
                        param,
                        "Generic param is missing a specialization in `#[axum_macros::debug_handler]`. Specify each generic param at least once using the 'with' attribute to support debugging generic functions.",
                    )
                );
            }
        }

        Ok(Specializer {
            generic_params,
            specializations,
        })
    }

    /// find which generic_param_idents are present in typ
    // fn find_generic_args_in_type<'a, 'b>(
    //     typ: &'a syn::Type,
    //     generic_param_idents: &'b HashSet<syn::Ident>,
    // ) -> HashSet<&'b syn::Ident> {

    // }

    /// Return vector of generic param identities found in the given type.
    ///
    /// Each param will be present in the returned vec at most once, and will be in order
    /// of appearance from self.generic_params
    fn find_generic_params<'a>(&'a self, typ: &'_ syn::Type) -> Vec<&'a syn::Ident> {
        let generic_param_set = HashSet::from_iter(self.generic_params.iter());
        let mut finder = GenericFinder {
            found_idents: HashSet::new(),
            generic_param_set,
        };
        finder.visit_type(typ);
        self.generic_params
            .iter()
            .filter(|i| finder.found_idents.contains(i))
            .collect()
    }

    fn compute_specializations<'a>(
        &'a self,
        typ: &'a syn::Type,
    ) -> Option<impl Iterator<Item = syn::Type> + 'a> {
        // to avoid generating the cross product of all generic params on
        // the function we first iterate over the type to find which generic
        // params are involed, then we generate the cross product
        // using the specializations of those params only.
        let ty_params = self.find_generic_params(typ);
        if ty_params.is_empty() {
            return None;
        }

        let ty_param_specializations = ty_params.iter().map(|param| {
            // safety: we can unwrap here due to the initializer invariant
            // that all generic params have at least one specialization
            self.specializations
                .get(param)
                .expect("there should be at least one specialization per generic type param")
                .iter()
        });

        Some(
            ty_param_specializations
                .multi_cartesian_product()
                .map(move |specializations| {
                    let param_specs: HashMap<&syn::Ident, &syn::Type> = HashMap::from_iter(
                        std::iter::zip(ty_params.iter().map(|f| *f), specializations),
                    );
                    let mut specializer = TypeSpecializer {
                        specializations: &param_specs,
                    };
                    let mut new_typ = typ.clone();
                    specializer.visit_type_mut(&mut new_typ);
                    new_typ
                }),
        )
    }

    /// For a given type, parameterized by the generics of item_fn, return all possible
    /// concrete types given the specified debug specializations. If the type does not
    /// contain any generic parameters, returns a vector with a single item which is a clone
    /// of the original type.
    ///
    /// Since some types may be generic over several params (eg `Foo<X, Y>`)
    /// the number of returned specializations is the size of the cross product of all
    /// specializations over the generic params that are present in the passed in typ.
    ///
    /// Substitution is done deeply, that is, for a given set of specializations the syntax tree
    /// for `typ` is searched deeply, recursively replacing each ocurrence of the generic parameters
    /// which ensures that all substitutions are made even in very complex cases such
    /// as `<<T as Trait>::Foo as some_crate::OtherTrait<U>>::Bar`.
    ///
    /// This function will only search for generic params named in `generic_params`, everything
    /// else will be assumed to be a concrete type.    
    ///
    /// Example:
    /// Assume a handler with two generic arguments `T` and `U` and the  debug specializations
    /// (T = u32, T = String, U = i32).
    ///     compute_all_specializations(Path<T>) would yield [Path<u32>, Path<String>]
    ///     compute_all_specializations(Path<U>) would yield [Path<i32>]
    ///     compute_all_specializations(Foo<T, U>) would yield [Foo<u32, i32>, Foo<String, i32>]
    ///     compute_all_specializations(U) would yield [i32]
    ///     compute_all_specializations(String) would yield [String]
    ///     
    pub(crate) fn all_specializations(&self, typ: &syn::Type) -> Vec<syn::Type> {
        self.compute_specializations(typ)
            .map(|i| i.collect())
            .unwrap_or_else(|| vec![typ.clone()])
    }

    /// Behaves the same way as `all_specializations` except only returns a single
    /// default specialization.
    pub(crate) fn specialize_default(&self, typ: &syn::Type) -> syn::Type {
        self.compute_specializations(typ)
            .and_then(|mut i| i.next())
            .unwrap_or_else(|| typ.clone())
    }

    /// Create a token stream with the default generic arg specializations
    /// for the handler, in turbofish syntax.
    ///
    /// For instance, for `fn handler<T, U>(t: Path<T>, u: Path<U>)` with `#[debug_handler(with(T = String, T = u64, U = i32))]`
    /// this would return `::<String, i32>`. The choice of the "default" specialization for each variable type
    /// is made by choosing the first specialization to appear in the `with` attribute.
    ///
    /// For a non-generic function, this returns the empty turbofish `::<>`.
    pub(crate) fn make_turbofish_with_default_specializations(&self) -> TokenStream {
        let default_handler_specializations = self.generic_params.iter().map(|f| {
            self.specializations
                .get(f)
                .and_then(|v| v.first())
                // safety: we can unwrap here due to the initializer invariant
                // that all generic params have at least one specialization
                .expect("there should be at least one specialization per generic type param")
        });
        quote! { ::<#(#default_handler_specializations),*> }
    }
}
