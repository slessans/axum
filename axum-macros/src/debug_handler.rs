use crate::{
    attr_parsing::second,
    with_position::{Position, WithPosition},
};
use proc_macro2::TokenStream;
use quote::{format_ident, quote, quote_spanned};
use syn::{parse_quote, spanned::Spanned, FnArg, ItemFn, Type};

use self::attr::Attrs;
use self::specializer::Specializer;

mod attr;
mod specializer;

pub(crate) fn expand(attr: Attrs, item_fn: ItemFn) -> TokenStream {
    let Attrs {
        body_ty,
        state_ty,
        with_tys,
    } = attr;
    let body_ty = body_ty
        .map(second)
        .unwrap_or_else(|| parse_quote!(axum::body::Body));
    let mut state_ty = state_ty.map(second);

    // these checks don't require the specializer, so we can generate code for them regardless
    // of whether we can successfully create one
    let check_extractor_count = check_extractor_count(&item_fn);
    let check_path_extractor = check_path_extractor(&item_fn);

    if state_ty.is_none() {
        state_ty = state_type_from_args(&item_fn);
    }
    let state_ty = state_ty.unwrap_or_else(|| syn::parse_quote!(()));

    // If the function is generic and an improper `with` statement was provided to the macro, we can't
    // reliably check its inputs or outputs. This will result in an error. We skip those checks to avoid
    // unhelpful additional compiler errors.
    let specializer_checks = match Specializer::new(with_tys, &item_fn) {
        Ok(specializer) => {
            let check_output_impls_into_response =
                check_output_impls_into_response(&item_fn, &specializer);
            let check_inputs_impls_from_request =
                check_inputs_impls_from_request(&item_fn, &body_ty, state_ty, &specializer);
            let check_future_send = check_future_send(&item_fn, &specializer);
            quote! {
                #check_output_impls_into_response
                #check_inputs_impls_from_request
                #check_future_send
            }
        }
        Err(err) => err.into_compile_error(),
    };

    quote! {
        #item_fn
        #check_extractor_count
        #check_path_extractor
        #specializer_checks
    }
}

fn check_extractor_count(item_fn: &ItemFn) -> Option<TokenStream> {
    let max_extractors = 16;
    if item_fn.sig.inputs.len() <= max_extractors {
        None
    } else {
        let error_message = format!(
            "Handlers cannot take more than {} arguments. \
            Use `(a, b): (ExtractorA, ExtractorA)` to further nest extractors",
            max_extractors,
        );
        let error = syn::Error::new_spanned(&item_fn.sig.inputs, error_message).to_compile_error();
        Some(error)
    }
}

fn extractor_idents(item_fn: &ItemFn) -> impl Iterator<Item = (usize, &syn::FnArg, &syn::Ident)> {
    item_fn
        .sig
        .inputs
        .iter()
        .enumerate()
        .filter_map(|(idx, fn_arg)| match fn_arg {
            FnArg::Receiver(_) => None,
            FnArg::Typed(pat_type) => {
                if let Type::Path(type_path) = &*pat_type.ty {
                    type_path
                        .path
                        .segments
                        .last()
                        .map(|segment| (idx, fn_arg, &segment.ident))
                } else {
                    None
                }
            }
        })
}

fn check_path_extractor(item_fn: &ItemFn) -> TokenStream {
    let path_extractors = extractor_idents(item_fn)
        .filter(|(_, _, ident)| *ident == "Path")
        .collect::<Vec<_>>();

    if path_extractors.len() > 1 {
        path_extractors
            .into_iter()
            .map(|(_, arg, _)| {
                syn::Error::new_spanned(
                    arg,
                    "Multiple parameters must be extracted with a tuple \
                    `Path<(_, _)>` or a struct `Path<YourParams>`, not by applying \
                    multiple `Path<_>` extractors",
                )
                .to_compile_error()
            })
            .collect()
    } else {
        quote! {}
    }
}

fn is_self_pat_type(typed: &syn::PatType) -> bool {
    let ident = if let syn::Pat::Ident(ident) = &*typed.pat {
        &ident.ident
    } else {
        return false;
    };

    ident == "self"
}

fn check_inputs_impls_from_request(
    item_fn: &ItemFn,
    body_ty: &Type,
    state_ty: Type,
    specializer: &Specializer,
) -> TokenStream {
    let takes_self = item_fn.sig.inputs.first().map_or(false, |arg| match arg {
        FnArg::Receiver(_) => true,
        FnArg::Typed(typed) => is_self_pat_type(typed),
    });

    WithPosition::new(item_fn.sig.inputs.iter())
        .enumerate()
        .map(|(arg_idx, arg)| {
            let must_impl_from_request_parts = match &arg {
                Position::First(_) | Position::Middle(_) => true,
                Position::Last(_) | Position::Only(_) => false,
            };

            let arg = arg.into_inner();

            let (span, ty) = match arg {
                FnArg::Receiver(receiver) => {
                    if receiver.reference.is_some() {
                        return syn::Error::new_spanned(
                            receiver,
                            "Handlers must only take owned values",
                        )
                        .into_compile_error();
                    }

                    let span = receiver.span();
                    (span, syn::parse_quote!(Self))
                }
                FnArg::Typed(typed) => {
                    let ty = &typed.ty;
                    let span = ty.span();

                    if is_self_pat_type(typed) {
                        (span, syn::parse_quote!(Self))
                    } else {
                        (span, ty.clone())
                    }
                }
            };
            specializer
                .all_specializations_of_type(&ty)
                .enumerate()
                .map(|(specialization_idx, specialized_ty)| {
                    let check_fn = format_ident!(
                        "__axum_macros_check_{}_{}_{}_from_request_check",
                        item_fn.sig.ident,
                        arg_idx,
                        specialization_idx,
                        span = span,
                    );

                    let call_check_fn = format_ident!(
                        "__axum_macros_check_{}_{}_{}_from_request_call_check",
                        item_fn.sig.ident,
                        arg_idx,
                        specialization_idx,
                        span = span,
                    );                    

                    let call_check_fn_body = if takes_self {
                        quote_spanned! {span=>
                            Self::#check_fn();
                        }
                    } else {
                        quote_spanned! {span=>
                            #check_fn();
                        }
                    };

                    let check_fn_generics = if must_impl_from_request_parts {
                        quote! {}
                    } else {
                        quote! { <M> }
                    };
        
                    let from_request_bound = if must_impl_from_request_parts {
                        quote! {
                            #specialized_ty: ::axum::extract::FromRequestParts<#state_ty> + Send
                        }
                    } else {
                        quote! {
                            #specialized_ty: ::axum::extract::FromRequest<#state_ty, #body_ty, M> + Send
                        }
                    };

                    quote_spanned! {span=>
                        #[allow(warnings)]
                        fn #check_fn #check_fn_generics()
                        where
                            #from_request_bound,
                        {}
        
                        // we have to call the function to actually trigger a compile error
                        // since the function is generic, just defining it is not enough
                        #[allow(warnings)]
                        fn #call_check_fn()
                        {
                            #call_check_fn_body
                        }
                    }
                })
                .collect::<TokenStream>()
        })
        .collect::<TokenStream>()
}

/// generates specialized calls to the handler fn with each possible specialization. The
/// argument for each function param is a panic. This can be used to check that the handler output
/// value. The handler will have the proper receiver prepended to it if necessary (for instance Self::).
fn generate_mock_handler_calls<'a>(
    item_fn: &'a ItemFn,
    specializer: &'a Specializer,
) -> impl Iterator<Item = TokenStream> + 'a {
    let span = item_fn.span();
    let handler_name = &item_fn.sig.ident;
    specializer
        .all_specializations_as_turbofish()
        .map(move |turbofish| {
            // generic panics to use as the value for each argument to the handler function
            let args = item_fn.sig.inputs.iter().map(|_| {
                quote_spanned! {span=> panic!() }
            });
            if let Some(receiver) = self_receiver(item_fn) {
                quote_spanned! {span=>
                    #receiver #handler_name #turbofish (#(#args),*)
                }
            } else {
                // we have to repeat the item_fn in here as a dirty hack to
                // get around the situation where the handler fn is an associated
                // fn with no receiver -- we have no way to detect to use Self:: here
                quote_spanned! {span=>
                    {
                        #item_fn
                        #handler_name #turbofish (#(#args),*)
                    }
                }
            }
        })
}

fn check_output_impls_into_response(item_fn: &ItemFn, specializer: &Specializer) -> TokenStream {
    let return_ty_span = match &item_fn.sig.output {
        syn::ReturnType::Default => return quote! {},
        syn::ReturnType::Type(_, ty) => ty,
    }
    .span();
    generate_mock_handler_calls(item_fn, specializer)
        .enumerate()
        .map(|(specialization_idx, handler_call)| {
            let name = format_ident!(
                "__axum_macros_check_{}_{}_into_response",
                item_fn.sig.ident,
                specialization_idx
            );
            quote_spanned! {return_ty_span=>
                #[allow(warnings)]
                async fn #name() {
                    let value = #handler_call.await;

                    fn check<T>(_: T)
                        where T: ::axum::response::IntoResponse
                    {}

                    check(value);
                }
            }
        })
        .collect::<TokenStream>()
}

fn check_future_send(item_fn: &ItemFn, specializer: &Specializer) -> TokenStream {
    if item_fn.sig.asyncness.is_none() {
        match &item_fn.sig.output {
            syn::ReturnType::Default => {
                return syn::Error::new_spanned(
                    &item_fn.sig.fn_token,
                    "Handlers must be `async fn`s",
                )
                .into_compile_error();
            }
            syn::ReturnType::Type(_, ty) => ty,
        };
    }
    let span = item_fn.span();
    generate_mock_handler_calls(item_fn, specializer)
        .enumerate()
        .map(|(specialization_idx, handler_call)| {
            let name = format_ident!(
                "__axum_macros_check_{}_{}_future",
                item_fn.sig.ident,
                specialization_idx
            );
            quote_spanned! {span=>
                #[allow(warnings)]
                fn #name() {
                    let future = #handler_call;
                    fn check<T>(_: T)
                        where T: ::std::future::Future + Send
                    {}
                    check(future);
                }
            }
        })
        .collect::<TokenStream>()
}

fn self_receiver(item_fn: &ItemFn) -> Option<TokenStream> {
    let takes_self = item_fn.sig.inputs.iter().any(|arg| match arg {
        FnArg::Receiver(_) => true,
        FnArg::Typed(typed) => is_self_pat_type(typed),
    });

    if takes_self {
        return Some(quote! { Self:: });
    }

    if let syn::ReturnType::Type(_, ty) = &item_fn.sig.output {
        if let syn::Type::Path(path) = &**ty {
            let segments = &path.path.segments;
            if segments.len() == 1 {
                if let Some(last) = segments.last() {
                    match &last.arguments {
                        syn::PathArguments::None if last.ident == "Self" => {
                            return Some(quote! { Self:: });
                        }
                        _ => {}
                    }
                }
            }
        }
    }

    None
}

/// Given a signature like
///
/// ```skip
/// #[debug_handler]
/// async fn handler(
///     _: axum::extract::State<AppState>,
///     _: State<AppState>,
/// ) {}
/// ```
///
/// This will extract `AppState`.
///
/// Returns `None` if there are no `State` args or multiple of different types.
fn state_type_from_args(item_fn: &ItemFn) -> Option<Type> {
    let types = item_fn
        .sig
        .inputs
        .iter()
        .filter_map(|input| match input {
            FnArg::Receiver(_) => None,
            FnArg::Typed(pat_type) => Some(pat_type),
        })
        .map(|pat_type| &*pat_type.ty);
    crate::infer_state_type(types)
}

#[test]
fn ui() {
    crate::run_ui_tests("debug_handler");
}
