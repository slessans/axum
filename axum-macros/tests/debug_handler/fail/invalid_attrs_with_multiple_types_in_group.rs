use axum_macros::debug_handler;

#[debug_handler(with(T = String, U = u64))]
async fn handler() {}

fn main() {}
