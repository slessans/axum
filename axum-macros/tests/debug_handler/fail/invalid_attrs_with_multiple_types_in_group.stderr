error: All argument names in a group must match. Groups should be separated by `;`. (e.g. `T = Foo, T = Bar; U = i64` is valid, but `T = Foo, U = Bar` is not).
 --> tests/debug_handler/fail/invalid_attrs_with_multiple_types_in_group.rs:3:34
  |
3 | #[debug_handler(with(T = String, U = u64))]
  |                                  ^
